<!DOCTYPE html>
<html class="notranslate" dir="ltr" lang="en" translate="no">
 <head>
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1" name="viewport"/>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <meta content="Here, I write about the distribution of amino acid types along protein sequences. To this end I analyze all protein sequences in the Uniprot database." name="description"/>
  <meta content="blog, uniprot, protein sequences, bioinformatics, biology, amino acids" name="keywords"/>
  <meta content="Semidan Robaina Estevez" name="author"/>
  <title>
   Semid&aacute;n: Frequency of amino acids along protein sequences
  </title>
  <link href="/imgs/my-logo.png" rel="icon" sizes="32x32" type="image/png"/>
  <link href="/css/master.css" rel="stylesheet"/>
  <link href="/css/blog.css" rel="stylesheet"/>
  <link href="/css/blog_entry.css" rel="stylesheet"/>
  <link href="/vendor/fontawesome_all.min.css" rel="stylesheet"/>
  <script src="https://unpkg.com/commentbox.io/dist/commentBox.min.js">
  </script>
  <!-- highlight.js -->
  <!-- styles loaded dynamically in settings.js -->
  <script src="/vendor/highlight_js/highlight.pack.js">
  </script>
  <script src="/vendor/jquery-3.4.1.min.js">
  </script>
  <script>
   $(function() {
      $("#navbar-container").load("/navbar.html");
    });
    $(function() {
      $("#footer").load("/footer.html");
    });
    let logo = new Image();
    logo.src = "/imgs/my-logo.png";
  </script>
  <script src="/js/transitions.js">
  </script>
  <script src="/js/settings.js">
  </script>
  <script src="/js/blog-entry.js">
  </script>
 </head>
 <body>
  <div id="page-container">
   <div id="content-wrap">
    <div id="navbar-container">
    </div>
    <div id="blog-wrap">
     <div id="tag-container">
      <div class="topictag Biology"></div>
      <div class="topictag Statistics"></div>
      <div class="topictag Python"></div>
     </div>
     <div id="blog-entry-header">
      <p id="entry-title">
       Frequency of amino acids along protein sequences
      </p>
      <p id="entry-date">
       30 September, 2020
      </p>
     </div>
     <article id="blog-content">
      <p>Some time ago I came across an interesting blog post by <a href="http://dclure.org/labs/distributions-of-words-27k-novels/">David McClure</a>, where he analyzes the frequency of words across narrative time for thousands of novels. He shows that some words tend to accumulate at the beginning of the novels, others do so at the end, and yet other ones occur with a more or less constant frequency across the novel. This finding makes sense. Certain words, such as descriptive words, tend to accumulate at the beginning because (usually) that's where all characters and scenarios are introduced, while other words that are linked to the story's resolution, such as <em>farewell</em>, <em>death</em> or <em>murderer</em> tend to accumulate at the end.</p>
<p>As I was reading the post, I couldn't help to think about protein sequences — an example of these weird connections by brain does sometimes. If we imagine a protein as a "novel" and a word as an amino acid, how would the distribution of amino acids across protein sequence look like if we were to analyze thousands of protein sequences at the same time? Well, I did not know, although I guessed that most likely the distributions wouldn't be uniform (amino acids, like words, have different properties depending on their chemical structures). So I went for it, downloaded the entire <a href="www.uniprot.com">Uniprot database</a> containing thousands of protein sequences of hundreds of different species, and did the analyzes. Specifically, I employed the fully (human) curated part of the <em>Uniparc</em> section, where sequences are grouped in taxonomic clusters: bacteria, archaea, plants, fungi, human, vertebrates, invertebrates, and viruses. I wrote a python parser to extract the sequences from the downloaded text files and performed the analyzes. In this post, I'll share the results with the world! But first, I feel like I should give a small intro to proteins and sequences to keep this post somewhat self-contained... Here we go!</p>
<h2>Brief introduction to proteins</h2>
<p>Proteins are amazing biomolecules, they fulfill numerous tasks within cells: from building structures, such as the cytoskeleton, to serving as biochemical reaction catalysts, replicate and transcribe DNA, synthesize other proteins, sense external and internal regulatory signals, regulate gene expression... to name just a few!</p>
<p>Proteins are constituted by (rather long) chains of <a href="https://en.wikipedia.org/wiki/Amino_acid">amino acids</a>. Amino acids are very cool, rather small, molecules that have a constant part, an amino and an acid group (yeah, hence the name), and a variable part which is chemically diverse and unique to each amino acid. Additionally, the sequence in which amino acids are located is encoded in specific regions of the DNA, the genes. There are 20 <a href="https://en.wikipedia.org/wiki/Proteinogenic_amino_acid">proteinogenic</a> amino acids, that is, amino acids that form proteins — well 20 plus two more that are not encoded in the <a href="https://en.wikipedia.org/wiki/Genetic_code">genetic code</a>.</p>
<p>Here is a colorful depiction* of all 20 proteinogenic amino acids encoded in the genetic code:</p>
<p><img class="jupyter-post-image" alt="aminoacids" src="/imgs/blog/frequency_of_aminoacids_along_protein_sequences/amino-acid-structures_med.jpeg"/></p>
<p>*Cornell, B. 2016. Amino Acids. [ONLINE] Available at http://ib.bioninja.com.au. [Accessed 18 October 2018]</p>
<p>The sequence in which the amino acids are arranged determines the physicochemical properties of a protein, and so its possible functions. Moreover, these amino acid chains are arranged in three-dimensional structures through a process called <a href="https://en.wikipedia.org/wiki/Protein_folding">protein folding</a>, and the three-dimensional arrangement is also determined by the sequence. Thus, it may happen that two amino acids positioned far away in the sequence are actually physically close in the three-dimensional structure. Also, some regions in the sequence may stay in the inner part of the protein while others may be located in the outside and hence face the surrounding medium.</p>
<p>Here is an example of the sequence of amino acids that constitutes a human protein, the enzyme <a href="https://en.wikipedia.org/wiki/Fructose-bisphosphate_aldolase">Fructose-bisphosphate aldolase</a>, which catalyzes a key reaction within the <a href="https://en.wikipedia.org/wiki/Glycolysis">glycolysis</a> — a metabolic pathway that breaks down glucose into smaller molecules, obtaining energy in the process. The sequence is 364 amino acids long and can be found in the <a href="https://www.uniprot.org/uniprot/P04075">Uniprot database</a> of protein sequences:</p>
<p><strong>start</strong></br>
MPYQYPALTPEQKKELSDIAHRIVAPGKGILAADESTGSIAKRLQSIGTENTEENRRFYR
QLLLTADDRVNPCIGGVILFHETLYQKADDGRPFPQVIKSKGGVVGIKVDKGVVPLAGTN
GETTTQGLDGLSERCAQYKKDGADFAKWRCVLKIGEHTPSALAIMENANVLARYASICQQ
NGIVPIVEPEILPDGDHDLKRCQYVTEKVLAAVYKALSDHHIYLEGTLLKPNMVTPGHAC
TQKFSHEEIAMATVTALRRTVPPAVTGITFLSGGQSEEEASINLNAINKCPLLKPWALTF
SYGRALQASALKAWGGKKENLKAAQEEYVKRALANSLACQGKYTPSGQAGAAASESLFVS
NHAY</br>
<strong>end</strong></p>
<p>Note that each amino acid is represented by a single letter. For instance, the first amino acid <em>M</em> corresponds to <em>Methionine</em> (names can be found in the picture above). As a side note, <em>Methionine</em> is the first amino acid in almost all proteins, as the DNA sequence that codes for <em>Methionine</em> serves as a start signal for the ribosome during <a href="https://en.wikipedia.org/wiki/Translation_(biology)">translation</a> — the translation of genetic information into a protein sequence.</p>
<h2>And now the results!</h2>
<p>Now that we briefly reviewed what proteins are, we are ready to crunch numbers and output some results. I wrote some python functions to read the downloaded txt files, extract all protein sequences in them, and compute the frequency of each amino acid along the protein sequence — the code can be found <a href="https://github.com/Robaina/UNIPROT">here</a>. To this end, first we need to partition each sequence into several bins, in this case, 30, and then count the occurrence of each amino acid in each bin (representing a fragment of the sequence). I'll repeat this process for all eight files corresponding to groups: <em>plants</em>, <em>human</em>, <em>bacteria</em>, <em>archaea</em>, <em>fungi</em>, <em>vertebrates</em>, <em>invertebrates</em> and <em>viruses</em>. I'll also add another group which is the average across all previous groups — depicted in back in the plots below.</p>
<pre><code class="python">from ParseUniprot import getProteinSequences, getAAFrequencies,
                         plotAAFrequencies, computeAverageAAFrequency
%matplotlib inline

fileNames = ['uniprot_sprot_plants.txt', 'uniprot_sprot_human.txt',
             'uniprot_sprot_bacteria.txt', 'uniprot_sprot_archaea.txt',
             'uniprot_sprot_fungi.txt', 'uniprot_sprot_vertebrates.txt',
             'uniprot_sprot_invertebrates.txt', 'uniprot_sprot_viruses.txt']

colors = {'archaea': 'red', 'bacteria': 'orange', 'fungi': 'yellow',
          'plants': 'green', 'human': 'cyan', 'vertebrates': 'blue',
          'invertebrates': 'violet', 'viruses': 'brown', 'average': 'black'}

workDir = 'C:\\Users\\tinta\\OneDrive\\Documents\\Projects\\UNIPROT\\'

amino_acid_frequencies = {}
database_length = {}
for file in fileNames:
    group = file.split('_')[-1].split('.txt')[0]
    proteins = getProteinSequences(workDir + file)
    database_length[group] = len(proteins)
    print(f'Evaluating {group} database, with {database_length[group]} sequences')
    amino_acid_frequencies[group] = getAAFrequencies(proteins, nsections=30)
# Add average frequencies
amino_acid_frequencies['average'] = computeAverageAAFrequency(amino_acid_frequencies)
</code></pre>
<pre><code>Evaluating plants database, with 41934 sequences
Evaluating human database, with 20360 sequences
Evaluating bacteria database, with 333552 sequences
Evaluating archaea database, with 19521 sequences
Evaluating fungi database, with 33465 sequences
Evaluating vertebrates database, with 18259 sequences
Evaluating invertebrates database, with 27200 sequences
Evaluating viruses database, with 16792 sequences
</code></pre>
<pre><code class="python">plotAAFrequencies(amino_acid_frequencies, colors=colors, nsections=30)
</code></pre>
<p><img class="jupyter-post-image" alt="png" src="/imgs/blog/frequency_of_aminoacids_along_protein_sequences/output_2_0.png"/></p>
<p>Wow! that's neat! isn't it? it does look like the distribution isn't uniform along protein sequences for most amino acids! The case of <em>Methionine</em> is spectacular, but remember that almost all protein sequences start with this amino acid, so, that was expected. However, there are other, more interesting, cases. For instance, some amino acids are not very well represented at the beginning of the sequences for most groups, e.g.: <em>Valine</em>, <em>Cysteine</em>, <em>Aspartic acid</em>, <em>Tyrosine</em>, <em>Histidine</em>. Some tend to be better represent towards the end, like <em>Arginine</em> and <em>Lysine</em> while other ones'frequency tends to decrease towards the end, e.g.: <em>Threonine</em>, <em>Leucine</em>, <em>Phenylalanine</em>, or <em>Proline</em>.</p>
<p>Why is this happening? Well, I don't have a clear answer. To begin with, I haven't conducted a proper statistical analysis of the distribution, the previous results were just first impressions... My guess is that it may have something to do with which parts of the protein are exposed to the (polar) cytoplasmic medium and which parts are exposed to the protein's interior or a (non-polar) cell membrane — would be the case of <a href="https://en.wikipedia.org/wiki/Transmembrane_protein">transmembrane proteins</a>.</p>
<p>Anyway, it was fun to play with the Uniprot database and to unveil these distributions along the protein sequence. Perhaps another time I'll conduct a proper analysis of these results. Or perhaps <strong>you</strong> will! Whether this is the case or not, please leave your impressions in the comments below!</p>
     </article>
    </div>
    <div id="blog-navigator">
    </div>
    <div id="suggested-readings">
     <h1 id="suggested-readings-title">
      Related posts:
     </h1>
     <div id="suggested-readings-grid">
     </div>
    </div>
    <!-- Blog comments (commentbox.io) -->
    <div class="commentbox"></div>
   </div>
   <footer id="footer">
   </footer>
  </div>
  <script src="/js/scrollToTop.js">
  </script>
  <div id="return-to-start-arrow" onclick="scrollToTop()">
  </div>
  <script type="text/x-mathjax-config">
   MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"] ],
          processEscapes: true
        },
        "HTML-CSS": {
          linebreaks: { automatic: true, width: "container" }
        }
      });
  </script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/latest.js?config=default">
  </script>
 </body>
</html>
